### PR TITLE
Implement BloomFilter class

### PR DESCRIPTION
**Pull request by milaGGL**
- [x] Implement Bloomfilter class
- [x] write validation test cases for Bloomfilter
- [x] add golden test

### CODE REVIEW

File: firebase-firestore/src/main/java/com/google/firebase/firestore/remote/BloomFilter.java
Comment (milaGGL) [inline firebase-firestore/src/main/java/com/google/firebase/firestore/remote/BloomFilter.java:122]: TODO: looking for a better way to keep only the lowest 64bit when overflow

File: firebase-firestore/src/main/java/com/google/firebase/firestore/remote/BloomFilter.java
Comment (dconeybe) [inline firebase-firestore/src/main/java/com/google/firebase/firestore/remote/BloomFilter.java:30]: nit: Remove `@NonNull` for primitive types (i.e. int, long, boolean, byte, float, double, char, short) because they are not nullable.

File: firebase-firestore/src/main/java/com/google/firebase/firestore/remote/BloomFilter.java
Comment (dconeybe) [inline firebase-firestore/src/main/java/com/google/firebase/firestore/remote/BloomFilter.java:56]: Add javadocs for getSize() and isEmpty(). The word "size" in this context is unclear: is it referring to the number of bytes in the bitmap or the number of bits of the bitmap that are part of the bloom filter? If possible, remove these methods altogether (or make them private) since I don't foresee them being useful outside of this class. If I'm wrong and they _are_ useful, then feel free to leave them as-is. If they are only used for unit testing, them give them default visibility (by removing the `public` visibilitly modifier) and add the `@VisibleForTesting` annotation.

File: firebase-firestore/src/main/java/com/google/firebase/firestore/remote/BloomFilter.java
Comment (dconeybe) [inline firebase-firestore/src/main/java/com/google/firebase/firestore/remote/BloomFilter.java:94]: Rather than returning null, re-throw the exception in a RuntimeException. `MessageDigest.getInstance("MD5")` should _never_ throw because Android always supports "MD5": https://developer.android.com/reference/java/security/MessageDigest. You should just be able to write `throw new RuntimeException(e);`.

File: firebase-firestore/src/main/java/com/google/firebase/firestore/remote/BloomFilter.java
Comment (dconeybe) [inline firebase-firestore/src/main/java/com/google/firebase/firestore/remote/BloomFilter.java:87]: nit: Add `@NonNull` to `MD5Hash()`. Also, rename it to `md5Hash` because it is unconventional to start Android method names with uppercase characters.

File: firebase-firestore/src/main/java/com/google/firebase/firestore/remote/BloomFilter.java
Comment (dconeybe) [inline firebase-firestore/src/main/java/com/google/firebase/firestore/remote/BloomFilter.java:55]: Remove `@NonNull` from `getSize()` since its return value is a primitive integer.

File: firebase-firestore/src/main/java/com/google/firebase/firestore/remote/BloomFilter.java
Comment (dconeybe) [inline firebase-firestore/src/main/java/com/google/firebase/firestore/remote/BloomFilter.java:90]: In the spirit of minimizing the amount of code in a try/catch block, move everything out of the try/catch block except the call that potentially throws the NoSuchAlgorithmException (i.e. the call to `MessageDigest.getInstance("MD5")`).

File: firebase-firestore/src/main/java/com/google/firebase/firestore/remote/BloomFilter.java
Comment (dconeybe) [inline firebase-firestore/src/main/java/com/google/firebase/firestore/remote/BloomFilter.java:71]: If `md5HashedValue.length != 16` then that is a critical, irrecoverable error. Throw a RuntimeException in this case with a message like `"invalid md5HashedValue.length: " + md5HashedValue.length + " (expected 16)"`.

File: firebase-firestore/src/main/java/com/google/firebase/firestore/remote/BloomFilter.java
Comment (dconeybe) [inline firebase-firestore/src/main/java/com/google/firebase/firestore/remote/BloomFilter.java:90]: I think you can just return digest.digest(value.getBytes()) and skip the call to update().

https://developer.android.com/reference/java/security/MessageDigest?hl=en#digest(byte[])

File: firebase-firestore/src/main/java/com/google/firebase/firestore/remote/BloomFilter.java
Comment (dconeybe) [inline firebase-firestore/src/main/java/com/google/firebase/firestore/remote/BloomFilter.java:109]: nit: Rename `num1` and `num2` to `hash1` and `hash2`, respectively.

File: firebase-firestore/src/main/java/com/google/firebase/firestore/remote/BloomFilter.java
Comment (dconeybe) [inline firebase-firestore/src/main/java/com/google/firebase/firestore/remote/BloomFilter.java:111]: nit: Use `long` instead of `Long` here. There is no need to create a boxed Long object.

Comment (dconeybe) [inline firebase-firestore/src/main/java/com/google/firebase/firestore/remote/BloomFilter.java:111]: nit: Add parentheses around `num2 * index` to be clear about the order of operations (if the linter lets you do this).

File: firebase-firestore/src/main/java/com/google/firebase/firestore/remote/BloomFilter.java
Comment (dconeybe) [inline firebase-firestore/src/main/java/com/google/firebase/firestore/remote/BloomFilter.java:129]: Maybe put bitmap at the _end_ of the String returned from toString() so that the hashCount and size are easy to find in the string (i.e. you don't have to scroll to the end of a potentially-long bitmap string).

File: firebase-firestore/src/main/java/com/google/firebase/firestore/remote/BloomFilter.java
Comment (dconeybe) [inline firebase-firestore/src/main/java/com/google/firebase/firestore/remote/BloomFilter.java:127]: Consider using Base64 to produce a more compact string representation that Arrays.toString().

https://developer.android.com/reference/android/util/Base64

File: firebase-firestore/src/test/java/com/google/firebase/firestore/remote/BloomFilterTest.java
Comment (dconeybe) [inline firebase-firestore/src/test/java/com/google/firebase/firestore/remote/BloomFilterTest.java:118]: Please find a way to remove this `@SuppressWarnings("DefaultCharset")` annotation. It should be pretty easy to fix by simply specifying "utf8" when decoding strings from bytes.

File: firebase-firestore/src/test/java/com/google/firebase/firestore/remote/BloomFilterTest.java
Comment (dconeybe) [inline firebase-firestore/src/test/java/com/google/firebase/firestore/remote/BloomFilterTest.java:124]: The golden files should be loaded from Android "resources", like how the spec test json files are.

The spec test json files are located in https://github.com/firebase/firebase-android-sdk/tree/master/firebase-firestore/src/test/resources/json. The directory name "json" is entirely arbitrary, so choose a different directory name for your golden test files.

File: firebase-firestore/src/test/java/com/google/firebase/firestore/remote/BloomFilterTest.java
Comment (dconeybe) [inline firebase-firestore/src/test/java/com/google/firebase/firestore/remote/BloomFilterTest.java:126]: nit: Use the full name "file" for the variable name, instead of "f". Non-standard abbreviations tend to detract from readability. In this case especially, saving 3 characters is not justified IMO.

File: firebase-firestore/src/test/java/com/google/firebase/firestore/remote/BloomFilterTest.java
Comment (dconeybe) [inline firebase-firestore/src/test/java/com/google/firebase/firestore/remote/BloomFilterTest.java:165]: Use Truth here instead of the base assertEquals. Using Truth I believe there is a way to attach some contextual information to the failure message. It would be really helpful if the failure message included the name of the golden file and the value of `i` that failed.

**Review by dconeybe — CHANGES_REQUESTED** [review]

File: firebase-firestore/src/test/java/com/google/firebase/firestore/remote/BloomFilterTest.java
Comment (milaGGL) [reply firebase-firestore/src/test/java/com/google/firebase/firestore/remote/BloomFilterTest.java:124]: The test cases are under Android "resources", under a directory named "bloom_filter_golden_test_data".

File: firebase-firestore/src/test/java/com/google/firebase/firestore/remote/BloomFilterTest.java
Comment (milaGGL) [reply firebase-firestore/src/test/java/com/google/firebase/firestore/remote/BloomFilterTest.java:165]: AssertEquals can take messages as well, and java keeps suggesting me simplify assertTrue with assertEquals.

File: firebase-firestore/src/main/java/com/google/firebase/firestore/remote/BloomFilter.java
Comment (milaGGL) [inline firebase-firestore/src/main/java/com/google/firebase/firestore/remote/BloomFilter.java:24]: could the size go over 2^32 bit? It happens when the bitmap length is over 536,870,912.

Comment (dconeybe) [reply firebase-firestore/src/main/java/com/google/firebase/firestore/remote/BloomFilter.java:24]: No. The size of the bitmap for 1000 entries is about 8000 bytes, which equals 64000 bits. The bitmap will never be large enough to overflow an int.

File: firebase-firestore/src/test/java/com/google/firebase/firestore/remote/BloomFilterTest.java
Comment (dconeybe) [reply firebase-firestore/src/test/java/com/google/firebase/firestore/remote/BloomFilterTest.java:123]: Okay I see that this is how the spec tests work too. Please ignore my comment. It was in error.

File: firebase-firestore/src/main/java/com/google/firebase/firestore/remote/BloomFilter.java
Comment (dconeybe) [inline firebase-firestore/src/main/java/com/google/firebase/firestore/remote/BloomFilter.java:53]: You can remove this javadoc comment for isEmpty(). Its name is self-explanatory. If you add back getSize() then that one should have a javadoc comment since "size" is ambiguous (number of bytes or number of bits).

File: firebase-firestore/src/test/java/com/google/firebase/firestore/remote/BloomFilterTest.java
Comment (dconeybe) [inline firebase-firestore/src/test/java/com/google/firebase/firestore/remote/BloomFilterTest.java:40]: Names of test methods in Java don't need to begin with "test" since they are annotated with `@Test`. Instead, please rename the methods to something like "mightContainShouldReturnFalseOnEmptyBloomFilter". Please rename the other test methods to follow this convention.

Comment (dconeybe) [inline firebase-firestore/src/test/java/com/google/firebase/firestore/remote/BloomFilterTest.java:40]: For an empty bloom filter, instead of checking isEmpty() (a private method) just call mightContain() with a bunch of different values, including the empty string, and expect it to return false every time.

File: firebase-firestore/src/test/java/com/google/firebase/firestore/remote/BloomFilterTest.java
Comment (dconeybe) [inline firebase-firestore/src/test/java/com/google/firebase/firestore/remote/BloomFilterTest.java:46]: An another example, a better name for "testEmptyBloomFilterThrowException" is "constructorShouldThrowIAEOnNegativeHashValue". Also, this test case should be split up so that each test case tests only one thing. The tests for different invalid arguments should be in different test methods.

go/unit-testing-practices#behavior-testing-examples

File: firebase-firestore/src/test/java/com/google/firebase/firestore/remote/BloomFilterTest.java
Comment (dconeybe) [inline firebase-firestore/src/test/java/com/google/firebase/firestore/remote/BloomFilterTest.java:119]: Make `documentPrefix` a private static final constant in this class, named something like `GOLDEN_DOCUMENT_PREFIX`.

Comment (dconeybe) [inline firebase-firestore/src/test/java/com/google/firebase/firestore/remote/BloomFilterTest.java:119]: I'd like to create one test method per golden test. Otherwise, the first golden test failure will hide any other failures from any other tests. I think the test way to do this would be to create a helper method like "runGoldenTest(String name)" and then a bunch of `@Test` methods that just call the runGoldenTest() method with the name of the test, which will be the file name.

**Review by dconeybe — CHANGES_REQUESTED** [review]

File: firebase-firestore/src/test/java/com/google/firebase/firestore/remote/BloomFilterTest.java
Comment (milaGGL) [reply firebase-firestore/src/test/java/com/google/firebase/firestore/remote/BloomFilterTest.java]: Strange, this is already removed in local and pushed up.

File: firebase-firestore/src/test/java/com/google/firebase/firestore/remote/BloomFilterTest.java
Comment (milaGGL) [reply firebase-firestore/src/test/java/com/google/firebase/firestore/remote/BloomFilterTest.java:133]: using InputStreamReader instead:
Screenshot 2023-01-09 at 3 19 12 PM

File: firebase-firestore/src/test/java/com/google/firebase/firestore/remote/BloomFilterTest.java
Comment (milaGGL) [reply firebase-firestore/src/test/java/com/google/firebase/firestore/remote/BloomFilterTest.java:40]: This mightContain on empty BloomFilter is tested separately below.  
Related test case: non empty string on empty bloom filter, empty string on empty bloom filter, empty string on non empty bloom filter. All should return false.

**Review by dconeybe — CHANGES_REQUESTED** [review]
I'm not quite done my review, but here's some feedback that I've accumulated so far.

File: firebase-firestore/src/main/java/com/google/firebase/firestore/remote/BloomFilter.java
Comment (milaGGL) [reply firebase-firestore/src/main/java/com/google/firebase/firestore/remote/BloomFilter.java:168]: Either way works for me.  But without the "", calling toString  will return this:
`BloomFilter{hashCount=1, size=7, bitmap=AQ==}`
In contrast:`BloomFilter{hashCount=1, size=7, bitmap="AQ=="}` looked prettier to me, especially with the "=" trailing at the end.

File: firebase-firestore/src/main/java/com/google/firebase/firestore/remote/BloomFilter.java
Comment (dconeybe) [inline firebase-firestore/src/main/java/com/google/firebase/firestore/remote/BloomFilter.java:70]: Add something like ", or false if the given string is definitely not contained in the bloom filter" after "true if the given string might be contained in the bloom filter".

I generally recommend to explicitly document the situations under which "true" and "false" are returned from a method, even if it's "obvious". It may not be so "obvious" to someone else.

File: firebase-firestore/src/main/java/com/google/firebase/firestore/remote/BloomFilter.java
Comment (dconeybe) [inline firebase-firestore/src/main/java/com/google/firebase/firestore/remote/BloomFilter.java:103]: nit: add the `static` modifier to `createMd5HashMessageDigest()`.

File: firebase-firestore/src/main/java/com/google/firebase/firestore/remote/BloomFilter.java
Comment (dconeybe) [inline firebase-firestore/src/main/java/com/google/firebase/firestore/remote/BloomFilter.java:105]: nit: Collapse the two spaces before `MessageDigest` to a single space.

File: firebase-firestore/src/main/java/com/google/firebase/firestore/remote/BloomFilter.java
Comment (dconeybe) [inline firebase-firestore/src/main/java/com/google/firebase/firestore/remote/BloomFilter.java:124]: Consider renaming the `index` argument to something like `hashIndex` or `hashNum`. There are two "indexes" being discussed in the documentation and IMO it's helpful to be clear which index is being referred to.

File: firebase-firestore/src/main/java/com/google/firebase/firestore/remote/BloomFilter.java
Comment (dconeybe) [inline firebase-firestore/src/main/java/com/google/firebase/firestore/remote/BloomFilter.java:152]: nit: Remove the parentheses around `(index / 8)` since they are superfluous.

File: firebase-firestore/src/test/java/com/google/firebase/firestore/remote/BloomFilterTest.java
Comment (dconeybe) [inline firebase-firestore/src/test/java/com/google/firebase/firestore/remote/BloomFilterTest.java:51]: For tests like `instantiateNonEmptyBloomFilter()`, which effectively test more than one thing, separate each test into its own logic scope using `{` and `}`. This provides several benefits, including

1. It logically groups the two tests together, improving readability.
2. It prevents accidental usage of variables from one test in another, improving robustness.
3. Variable names can be re-used, increasing the symmetry between the tests, making it easier for a human to see what's the same and what's different between the cases.

For example, the body of `instantiateNonEmptyBloomFilter()` would become this:

```java
{
  BloomFilter bloomFilter = new BloomFilter(new byte[] {1}, 0, 1);
  assertEquals(bloomFilter.getBitCount(), 8);
}
{
  BloomFilter bloomFilter = new BloomFilter(new byte[] {1}, 7, 1);
  assertEquals(bloomFilter.getBitCount(), 1);
}
```

File: firebase-firestore/src/main/java/com/google/firebase/firestore/remote/BloomFilter.java
Comment (dconeybe) [inline firebase-firestore/src/main/java/com/google/firebase/firestore/remote/BloomFilter.java:50]: Suggested re-wording of this exception's message:

```
"Invalid padding when bitmap length is 0: " + padding
```

to

```
"Expected padding of 0 when bitmap length is 0, but got " + padding
```

The re-wording is a bit better because it also conveys the _expected_ value, not just the _invalid_ value.

File: firebase-firestore/src/test/java/com/google/firebase/firestore/remote/BloomFilterTest.java
Comment (dconeybe) [inline firebase-firestore/src/test/java/com/google/firebase/firestore/remote/BloomFilterTest.java:102]: nit: Suggest renaming `constructorShouldThrowIAEOnOverflowPadding()` to `constructorShouldThrowIAEIfPaddingIsTooLarge()`. To me, the term "overflow" (incorrectly) implies that the value was once valid but grew to size that overflowed the valid range.

File: firebase-firestore/src/test/java/com/google/firebase/firestore/remote/BloomFilterTest.java
Comment (dconeybe) [inline firebase-firestore/src/test/java/com/google/firebase/firestore/remote/BloomFilterTest.java:151]: Verify that `resultFile.equals(testFile) == false`. If the `testFile` does not contain "bloom_filter_proto" then both `testFile` and `resultFile` will be equal, which is not right. Throw IllegalArgumentException in this case.

File: firebase-firestore/src/test/java/com/google/firebase/firestore/remote/BloomFilterTest.java
Comment (dconeybe) [inline firebase-firestore/src/test/java/com/google/firebase/firestore/remote/BloomFilterTest.java:171]: You can remove the `testFile` from the failure message, since the test file can now be inferred from the test case that failed.

File: firebase-firestore/src/test/java/com/google/firebase/firestore/remote/BloomFilterTest.java
Comment (dconeybe) [inline firebase-firestore/src/test/java/com/google/firebase/firestore/remote/BloomFilterTest.java:170]: Consider this re-wording of the failure message: `mightContain(...) returned X, but expected Y`.

File: firebase-firestore/src/test/java/com/google/firebase/firestore/remote/BloomFilterTest.java
Comment (dconeybe) [inline firebase-firestore/src/test/java/com/google/firebase/firestore/remote/BloomFilterTest.java:181]: This comment basically repeats the method's name. Just remove it.

File: firebase-firestore/src/test/java/com/google/firebase/firestore/remote/BloomFilterTest.java
Comment (dconeybe) [inline firebase-firestore/src/test/java/com/google/firebase/firestore/remote/BloomFilterTest.java:150]: Mark `runGoldenTest()` and `readJsonFile()` as `static`, if possible.

**Review by dconeybe — CHANGES_REQUESTED** [review]
All of my remaining comments are minor nits. Thanks!

File: firebase-firestore/src/test/java/com/google/firebase/firestore/remote/BloomFilterTest.java
Comment (dconeybe) [reply firebase-firestore/src/test/java/com/google/firebase/firestore/remote/BloomFilterTest.java]: Please make this change in `constructorShouldThrowNPEOnNullBitmap()`, `constructorShouldThrowIAEOnNegativePadding()`, `constructorShouldThrowIAEOnNegativeHashCount()`, `mightContainWithEmptyStringMightReturnFalsePositiveResult()`, and `bloomFilterToString()` too.

**Review by dconeybe — CHANGES_REQUESTED** [review]
Just one very minor nit, then ready to merge! Feel free to merge without waiting for my final approval since I may be OOO when you address that comment.

**Review by dconeybe — APPROVED** [review]

**Pull request merged**
